File: topological sort-dfs.cpp
// topological sort is only possible in DAG
class Solution
{
	public:
	void dfs(int current, vector<int> adj[], vector<int> &visited, vector<int> &stack){
	    visited[current] = 1;
	    for (auto &adjacentNode: adj[current]){
	        if (!visited[adjacentNode]){
	            dfs(adjacentNode, adj, visited, stack);
	        }
	    }
	    stack.push_back(current);
	}
	//Function to return list containing vertices in Topological order. 
	vector<int> topoSort(int V, vector<int> adj[]) 
	{
	    // code here
	    vector<int> visited(V, 0);
	    vector<int> stack;
	    
	    for (int at = 0; at < V; at++){
	        if (!visited[at]){
	            dfs(at, adj, visited, stack);
	        }
	    }
	    reverse(stack.begin(),stack.end());
	    return stack;
	}
};

// TC: O(V + E)

File: topological sort-dfs.py
class Solution:
    def dfs(self, current, visited, adj, stack):
        visited[current] = 1
        for adjacentNode in adj[current]:
            if not visited[adjacentNode]:
                self.dfs(adjacentNode, visited, adj, stack)
                
        # by recursively,  we have already inserted nodes into stack that comes after current node
        # which ensures that U comes before V if U is directed to V
        # Now insert the current node into the stack
        stack.append(current)
        
        
    #Function to return list containing vertices in Topological order.
    def topoSort(self, V, adj):
        # Code here
        visited = [0 for _ in range(V)]
        stack = []
        for i in range(V):
            if  not visited[i]:
                self.dfs(i, visited, adj, stack);
     
                
        return stack[::-1]

# TC: O(V) + O(E) Space: O(V)
